<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>FIMO - a motif search tool</title>
    <link href="./meme-suite.css" rel="stylesheet" type="text/css" />
  </head>
  <body>
    <div id="main">
      <script src="./meme-suite-logo.js" type="text/javascript">&gt;</script>
      <img src="images/fimo_logo.png" alt="FIMO logo" />
      <p><b>Usage:</b></p>
      <p>
      <code>fimo [options] &lt;motifs&gt; &lt;database&gt;</code>
      </p>
      <p>
      <b>Description:</b>
      </p>
      <p>
      The name <code>FIMO</code> stands for "Find Individual Motif Occurences."  
      The program searches a sequence database for occurrences of known motifs,
      treating each motif independently.  
      The program uses a dynamic programming algorithm to convert log-odds scores (in bits) into p-values,
      assuming a zero-order background model.
      By default the program reports all motif occurrences with a p-value less than <code>1e-4</code>.
      The threshold can be set using the <code>--thresh</code> option.
      The p-values for each motif occurence are converted to q-values following the method of
      Benjamini and Hochberg ("q-value" is defined as the minimal
      false discovery rate at which a given motif occurrence is deemed significant).
      The <code>--qv-thresh</code> option directs the program to use q-values rather than p-values
      for the threshold.
      If a motif has the <code>strand</code> feature set
      to <code>+/-</code> (rather than <code>+</code>), then
      <code>fimo</code> will search both strands for occurrences.
      </p>
      <p>
      The parameter <code>--max-stored-scores</code> sets the maximum number of motif
      occurrences that will be retained in memory.
      It defaults to 100,000.
      If the number of matches found reaches the maximum value allowed, 
      FIMO will discard 50% of the least significant matches,
      and new matches falling below the significance level of the retained matches
      will also be discarded.
      </p>
      <p>
      <code>FIMO</code> can make use of position specific priors (PSP) to improve its
      identification of true motif occurrences.
      To take advantage of PSP in FIMO you use must provide two command line options.
      The <code>--psp</code> option is  used to set the name of a 
      <a href="meme.html#MEME-PSP">MEME PSP file</a>,
      and the <code>--prior-dist</code> option is used to set the name of a file
      containing the binned distribution of priors.
      </p>

      <p><b>Input:</b></p> 
      <ul> 
      <li>
      <code>&lt;motifs&gt;</code> is the name of a file containing 
      a list of motifs, in 
      <a href="meme-format.html">MEME format</a>.
      </li>
      <li>
      <code>&lt;database&gt;</code> is the name of a file containing
      a collection of sequences in
      <a href="fasta-format.html">FASTA</a> format. 
      The character <code>-</code> can be used to indicate that the sequence data
      should be read from standard input.
      This can only be used if the motif file contains a single motif.
      </li>
      </ul>
      <p>
      The FASTA header lines are used as the source of sequence names.
      The sequence name is the string following the initial '&gt;' up
      to the first white space character.
      If the sequence name is of the form: <code>text:number-number</code>,
      the text portion will be used as the sequence name.
      The numbers will be used as genomic coordinates,
      and the first number will be used as the coordinate of the first position of the sequence.
      In all other cases the coordinate of the first postion of the sequence is
      take as 1.
      </p>
      <p><b>Output:</b></p>
      <p>
      FIMO will create a directory, named <code>fimo_out</code> by default.
      Any existing output files in the directory will be overwritten.
      The directory will contain:
      </p>
      <ul>
      <li>
      An XML file named <code>fimo.xml</code> using the 
      <a href="http://zlab.bu.edu/~phaverty/SupplementalData/CisML">CisML</a> schema.
      </li>
      <li>
      An HTML file named <code>fimo.html</code>
      </li>
      <li>
      A plain text file named <code>fimo.text</code>
      </li>
      <li>
      A plain text file in 
      <a href="http://www.sanger.ac.uk/resources/software/gff/">GFF format</a>
      named <code>fimo.gff</code>
      </li>
      </ul>
      <p>
      The default output directory can be overridden using the <code>--o</code> or <code>--oc</code>
      options which are described below.
      </p>
      <p>
      The <code>--text</code> option will limit output to plain text sent to the standard output.
      This will diable the calcualtion of q-values.
      </p>
      <p>
      The score reported in the GFF output is <code>min(1000, -10*(log10(pvalue)))</code>.
      </p>
      <p>The HTML and plain text output contain the following columns:</p>
      <ul>
        <li>The motif identifier</li>
        <li>The sequence identiifer</li>
        <li>The strand
        <code>+</code> indicates the motif matched the forward strand,
        <code>-</code> the reverse strand,
        and <code>.</code> indicates strand is not applicable (as for amino acid sequences).</li>
        <li>The start position of the motif occurence (closed, 1-based coordinates, unless genomic coordinates are provided)</li>
        <li>The end position of the motif occurence (closed, 1-based coordinates, unless genomic coordinates are provided). 
        </li>
        <li>The score for the motif occurence.
        The score is computed by by summing the appropriate entries from each column of
        the position-dependent scoring matrix that represents the motif.</li>
        <li>The p-value of the motif occurence. 
        The p-value is the
        probability of a random sequence of the same length as the motif
        matching that position of the sequence with a score at least as good.
        </li>
        <li>The q-value of the motif occurence.
        The q-value is the estimated
        false discovery rate if the occurrence is accepted as significant.
        See Storey JD, Tibshirani R. Statistical significance for genome-wide studies. 
        <i>Proc. Natl Acad. Sci. USA (2003) 100:9440â€“9445</i>
        </li>
        <li>
        The sequence matched to the motif.
        </li>
      </ul>
      <p>The HTML and plain text output is sorted by increasing p-value.</p>
      <p>
      <b>Options:</b> 
      </p>
      <ul>
      <li><code>--alpha <i>&lt;float&gt;</i></code>
    - The alpha parameter for calculating position specific priors.
    Alpha represents the fraction of all transcription factor binding 
    sites that are binding sites for the TF of interest.
    Alpha must be between 0 and 1.
    The default value is 1.0.
  </li>
      <li><code>--bgfile <i>&lt;bfile&gt;</i></code> 
    - Read background frequencies from
    <code><i>&lt;bfile&gt;</i></code>.  
    The file should be in 
    <a href="bfile-format.html">MEME background file format</a>.
    The default is to use frequencies
    embedded in the application from the non-redundant database.  If the
    argument is the keyword <code>motif-file</code>, then the frequencies
    will be taken from the motif file.
  </li>
      <li><code>--max-seq-length <code><i>&lt;max&gt;</i></code></code>
    - Set the maximum length allowed for input sequences.
    By default the maximum allowed length is 250000000.
  </li>
      <li><code>--max-strand</code>
    - If matches on both strands at a given position satisfy the output threshold,
    only report the match for the strand with the higher score.
    By default both matches are reported.
  </li>
      <li><code>--max-stored-scores <code><i>&lt;max&gt;</i></code></code>
    - Set the maximum number of scores that will be stored.
    Precise calculation of q-values depends on having a complete list of scores.
    However, keeping a complete list of scores may exceed available memory.
    Once the number of stored scores reaches the maximum allowed, 
    the least significant 50% of scores will be dropped,
    and approximate q-values will be calculated.
    By default the maximum number of stored matches is 100,000.
  </li>
      <li><code>--motif <i>&lt;id&gt;</i></code>
    - Use only the motif identified by <code><i>&lt;id&gt;</i></code>. 
    This option may be repeated.
  </li>
      <li><code>--motif-pseudo <i>&lt;float&gt;</i></code>
    - A pseudocount to be added to each count in the motif matrix,
    after first multiplying by the corresponding background frequency (default=0.1).
  </li>
      <li><code>--no-qvalue</code>
    - Do not compute a q-value for each p-value.  The q-value 
    calculation is that of Benjamini and Hochberg (1995).
    By default, q-values are computed.
  </li>
      <li><code>--norc</code>
    - Do not score the reverse complement DNA strand.  Both
	strands are scored by default.
  </li>
      <li><code>--o <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will not be overwritten.
  </li>
      <li><code>--oc <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will be overwritten.
  </li>
      <li><code>--parse-genomic-coord</code>
    When this option is specified each sequence header will be checked
    for UCSC style genomic coordinates. 
    These are of the form
    <pre>
      &gt;<i>seq-name</i>:<i>starting pos.</i>-<i>ending pos.</i></pre>
    where <i>seq-name</i> is the name of the sequence,
    <i>starting pos.</i> is the index of the first base,
    and 
    <i>ending pos.</i> is the index of the final base.
    <i>seq-name</i> may not contain any white space.
    If genomic coordinates are found they will be used
    as the coordinates in the output.
    If no coordinate are found, the first position in the 
    sequence will assumed to be 1.
  </li>
      <li><code>--psp <i>&lt;file&gt;</i></code>
    - File containing position specific priors (PSP) in
    <a href="meme.html#MEME-PSP">MEME PSP format</a>.
  </li>
      <li><code>--prior-dist <i>&lt;file&gt;</i></code>
    - File containing binned distribution of priors.
    This file can be generated from a 
    <a href="meme.html#MEME-PSP">MEME PSP format file</a>.
    using the
    <a href="compute-prior-dist.html">compute-prior-dist</a> utility.
  </li>
      <li><code>--qv-thresh</code>
    - Directs the program to use q-values for the output threshold.
    The default is to use p-values.
  </li>
      <li><code>--text</code>
    Limits output to plain text sent to standard out.  For FIMO, the
    text output is unsorted, and q-values are not reported.  This mode
    allows the program to search an arbitrarily large database,
    because results are not stored in memory.
  </li>
      <li><code>--thresh<i>&lt;float&gt;</i></code>
    - The output threshold for displaying search results.
    Only search results with a p-value less than the threshold will be output.
    The default threshold is a p-value of 1e-4.
    The threshold can be set to use q-values rather than p-values via the <code>--qv-thresh</code> option.
  </li>
      <li><code>--verbosity 1|2|3|4</code>
    - Set the verbosity of status reports to standard error.
    The default level is 2.
  </li>
      </ul>
    </div>
  </body>
</html>
