<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
	<head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
		<title>AME - a motif enrichment analysis tool</title>
		<link href="./meme-suite.css" rel="stylesheet" type="text/css" />
	</head>
	<body>
		<div id="main">
			<script src="./meme-suite-logo.js" type="text/javascript">&gt;</script>
			<p><b>Usage:</b></p>
      <p>
			<code>ame [options] &lt;sequence file&gt; &lt;motif file&gt;+</code>
    	</p>
      <p>
      <b>Description:</b>
      </p>

      <p>
        <code>AME</code> (Analysis of Motif Enrichment)
	scores a set of DNA sequences given a set of DNA-binding motifs,
	treating each position in the sequence as a the starting point of a
	possible binding event. AME supports a wide variety of methods for
	scoring motif enrichment, many methods of testing the scored motif enrichment for
	significance. By default, AME counts the number of cases where the
	<i>p</i>-value of the binding event for each motif is below a given
	threshold, and performs
	a Fisher exact test versus the number of binding events in a background
	sequence set to determine the <i>p</i>-value of the count for each
	motif. The background set is appended to the main sequence set in the
	input file, and the offset within the file where the background starts
	is specified on the command line.
      </p>
      <p><b>Input:</b></p> 
      <ul> 
      <li>
      <code>&lt;sequence file&gt;</code> is a collection of sequences in
      <a href="fasta-format.html">FASTA</a> format. 
      </li>
      <li>
      <code>&lt;motif file&gt;</code> containing a list of motifs, in 
      <a href="meme-format.html">MEME format</a>. More than one file
      can be specified.
      </li>
      </ul>
      
    	<p><b>Output:</b></p>
      <p>
      <code>AME</code> writes to a directory, <code>ame_out</code>, unless a different
      directory name is specified on the command line. The output directory contains outputs
      in two formats: HTML and plain text, in files named respectively <code>ame.html</code> and
      <code>ame.txt</code>.
      </p>
      <p>
      <b>Options:</b> 
      </p>
      <ul>
    		<li><code>--o <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will not be overwritten.
  </li>
    		<li><code>--oc <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will be overwritten.
  </li>
    		<li><code>--method<i>&lt;fisher|mhg|4dmhg|ranksum|linreg|spearman&gt;</i></code> 
    - Select the association function for testing motif enrichment significance.
    Note that linear regression and spearman rank correlation tests do not calculate
    p-values. Please use RAMEN if you desire to use linear regression with <i>p</i>-values.
  </li>
    		<li><code>--scoring<i>&lt;avg|max|sum|totalhits&gt;</i></code> 
    - Method of scoring motif enrichment. Either average-odds, the maximum (single-site) odds,
    the sum of odds, or counting individual hits that are more significant than a threshold.
    If using <code>totalhits</code> mode, please see also<code>--pvalue-threshold</code></li>
    		<li><code>--bgformat 0..2</code> - Source for determining background frequencies<br /><p>
    0: uniform background<br />
    1: MEME motif file<br />
    2: Background file
  </p></li>
    		<li><code>--bgfile <i>&lt;bfile&gt;</i></code> 
    - Read background frequencies from
    <code><i>&lt;bfile&gt;</i></code>.  
    The file should be in 
    <a href="bfile-format.html">MEME background file format</a>.
    The default is to use frequencies
    based on the motif file or files. See also <code>--bgformat</code></li>
    		<li><code>--length-correction</code> - Correct for length bias: subtract expected hits.
  Default=no length correction.
  </li>
    		<li><code>--pvalue-threshold &lt;float, default=2e-4&gt;</code> - Threshold to consider single
  motif hit significant.
  </li>
    		<li><code>--fix-partition</code> - Number of positive sequences; the balance are
  used as the background.
  </li>
    		<li><code>--pvalue-report-threshold &lt;float, default=1e-3&gt;</code> - Corrected
  <i>p</i>-value threshold for reporting a motif.
  </li>
    		<li><code>--rsmethod<i>&lt;better|quick&gt;</i></code> 
    - Select whether to use a proper ranksum test (<code>better</code>) or a faster
    heuristic. Default is to use the proper test.
  </li>
    		<li><code>--poslist<i>&lt;fl|pwm&gt;</i></code> 
    - For partition maximization, test thresholds on either X (pwm) or Y (fluorescence score).
    Default is fluorescence score.
    Only applies for partition maximisation and for the Ranksum test.
  </li>
    		<li><code>--log-fscores</code> 
    - For linear regression and spearman tests only: regress using
    ln(fluorescence score), rather than the score directly.
  </li>
    		<li><code>--log-pwmscores</code> 
    - For linear regression and spearman tests only: regress using
    ln(pwm score), rather than the score directly.
  </li>
    		<li><code>--normalise-affinity</code> 
    - Normalise motif scores so that motif scores can be compared directly. Only relevant for
    Spearman and Linear Regression tests, where p-values are not calculated.
  </li>
    		<li><code>--linreg-switchxy</code> 
    - Make the x-points fluorescence scores and the y-points PWM scores.
    Only relevant for Spearman and Linear Regression tests.
  </li>
    		<li><code>--fl-threshold<i>&lt;p-value&gt;</i></code> 
    - Only used if <code>--poslist</code> is in use. <i>Maximum</i> fluorescence
    p-value to consider as a 'positive' when labelling positives.
    Default is 1e-3.
    Use for Fisher Test with either SUM, AVG or MAX scoring only.
    </li>
		<li><code>--pwm-threshold<i>&lt;score&gt;</i></code> 
    - Minimum PWM score to call a sequence a 'positive'. Default is 1.
    Use for Fisher Test with either SUM, AVG or MAX scoring only.
  </li>
		<li><code>--verbose &lt;1...5&gt;</code> - Integer describing verbosity (low number is
  less verbose). Best placed first.
  </li>
    		<li><code>--help</code> 
    - Print a usage statement.
  </li>
   		</ul>
   		<p>
   		The default output directory is <code>ame_out</code>, and is
   		not overwritten (i.e., the default is the same as <code>--o ame_out</code>).
   		</p>

   <h3>Notes</h3>
   <p>
	   This version of AME has different default settings to the original AME. 
	   By default, it will perform the Fisher exact test, and count individual 
	   motif hits (total hits mode). It will perform partition maximisation by
	   default, but using a fixed partition between foreground and background
	   sequences is likely to the be the preferred mode of operation.
     <code>fix-partition</code> in this version
     is optional, rather than required.
   </p>

   <h3>Citing ame</h3>
   <p>
     If AME is of use to you in your research, please cite:
   </p>
   <blockquote>
          Robert C. McLeay, Timothy L. Bailey (2009).<br />
          "Motif Enrichment Analysis: a unified framework and an evaluation
          on ChIP data."<br />
          <i>BMC Bioinformatics</i> 2010, 11:165, doi:10.1186/1471-2105-11-165.
   </blockquote>
   <h3>Contact the authors</h3>
   <p>
     You can contact the authors via email:
   </p>
   <blockquote>
         Robert McLeay <a href="mailto:r.mcleay@imb.uq.edu.au">r.mcleay@imb.uq.edu.au</a>, and
         Timothy Bailey <a href="mailto:t.bailey@imb.uq.edu.au">t.bailey@imb.uq.edu.au</a>.
   </blockquote>
   <p>
     Bug reports should be directed to Robert McLeay.
   </p>
		</div>
  </body>
</html>
