<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
    <title>motiph</title>
    <link href="./meme-suite.css" rel="stylesheet" type="text/css" />
  </head>
  <body>
    <div id="main">
      <script src="./meme-suite-logo.js" type="text/javascript">&gt;</script>
      <blockquote>
      <h1><img height="81" width="444" src="meta-meme.gif" /></h1>
      <center>
        <h1>motiph</h1>
      </center>

      <br />
      <b>Usage:</b> 
      <blockquote>
        <code>motiph [options] &lt;alignment file&gt; &lt;tree file&gt; &lt;motif file&gt;</code>
      </blockquote>
      <p>
        <b>Description:</b>
      </p>

      <blockquote>
      <p>
      Scan a DNA multiple sequence alignment for occurrences of given motifs,
      taking into account the phylogenetic tree relating the sequences.
      <b>Note:</b> any sequences containing nothing but gaps are removed from the
      alignment.
      </p>
      <p>
      The algorithm is as follows:
      </p>
      <pre>
      A = a multiple alignment
      T = a phylogenetic tree
      M = a motif frequency matrix
      B = the background frequencies of the four bases in A
      U = a background evolutionary model with equilibrium frequencies B

      // Build evolutionary models.
      for j in positions of M {
        E[j] = an evolutionary model with equilibrium frequencies 
          from the jth position of M
      }

      // Build matrix of all possible scores
      for i in all possible alignment columns {
        for j in positions of M {
          foreground = site_likelihood(E[j], A[:][i], T)
          background = site_likelihood(U, A[:][i] T)
          score_matrix[i][j] = log_2(foreground / background)
        }    
      }
      // Calculate p-values of all possible scores for motif sized windows
      // windows in the alignment.
      pvalues = get_pv_lookup(score_matrix, B)

      // Calculate score for each motif sized window in the alignment.
      for i in columns of A {
        score = 0
        for j in positions of M {
          index = calculate the index of A[:][i + j] in the array
            of all possible alignment columns
          score = score + score_matrix[index][j]
        }    
        print pvalues[score]
      }
      </pre>

      <p>
      The core of the algorithm is a routine (<code>site_likelihood</code>)
      for scoring a particular column of the multiple alignment using a
      given evolutionary model and a given phylogenetic tree.  The alignment
      site provides the observed nucleotides at the base of the tree, and we
      sum over the unobserved nucleotides in the rest of the tree,
      conditioning on the equilibrium distribution from the evolutionary
      model at the root of the tree (Felsenstein Pruning Algorithm).  The
      tree must be a maximum likelihood tree, of the kind generated by DNAml
      from Phylip or by FastDNAml.  Branch lengths in the tree are converted
      to conditional probabilities using the specified evolutionary
      model.</p>

      </blockquote>

      <b>Input:</b> 
      <ul> 
      <li>
      <code>&lt;alignment file&gt;</code> is a DNA multiple alignment in 
      <a href="clustalw-format.html">ClustalW</a> format. 
      Alternatively, if the <code>--list</code> option is used,
      this file may contain a list of alignment files.
      In this case each of the alignment files will be scanned for
      each of the motifs.</li>

      <li>
      <code>&lt;tree file&gt;</code> is a phylogenetic tree in 
      <a href="http://evolution.genetics.washington.edu/phylip/newicktree.html">
        Phylip Newick
      </a> 
      format.  This tree may contain
      additional species not represented in the alignment.</li>

      <li>
        <code>&lt;motif file&gt;</code> is a list of motifs, in 
        <a href="meme-format.html">MEME format</a>.
      </li>

      </ul>

      <b>Output:</b>

      <blockquote>
      <p>
      The output is in 
      <a href="http://www.sanger.ac.uk/Software/formats/GFF/">GFF</a> 
      format with the forward and reverse scores listed as separate features.
      The scores are the <i>p</i>-values for the observed alignment columns in the motif sized windows.
      </p>
      </blockquote>

      <b>Options:</b>
      <ul>
      <li> <code>--bg <i>&lt;float&gt;</i></code>
    - The mutation rate for sites in the
    background model.  The default value is 1.
  </li>
      <li> <code>--bgfile <i>&lt;bfile&gt;</i></code> 
    - Read background frequencies from
    <code><i>&lt;bfile&gt;</i></code>.  
    The file should be in 
    <a href="bfile-format.html">MEME background file format</a>.
    The default is to use frequencies
    embedded in the application from the non-redundant database.  If the
    argument is the keyword <code>motif-file</code>, then the frequencies
    will be taken from the motif file.
  </li>
      <li> <code>--column-freqs [simulated|empirical]</code>
    - The way to compute the frequencies of all possible columns.
    <ul><li><code>simulated</code> - Use the evolutionary model to compute the frequency of each
      possible column of letters.</li><li><code>empirical</code> - Count the numbers of each column in the input multiple
      alignments.  All alignments using the same (sub)set of species are 
      counted together.  Frequencies are computed by dividing each by the 
      total counts for that (sub)set of species.</li></ul>
    The default is <code>simulated</code>.  These frequencies are used for
    computing <i>p</i>-values for scores.  If <code>simulated</code> is used, 
    the accuracy of the <i>p</i>-values depends strongly on the accuracy of the 
    evolutionary model.
  </li>
      <li> <code>--fg <i>&lt;float&gt;</i></code>
    - The mutation rate for sites in the foreground model(s).
    The default value is 1.
  </li>
      <li> <code>--gap <i>&lt;method&gt;</i></code>
    - Specifies the gap handling strategy.
    Allowed values for method are:
    <ul><li><code>skip</code> Skip those sites where any position in the alignment
      window contains a gap. This is the default gap handling strategy.
      </li><li><code>fixed</code> Sites containing gaps are assigned a fixed
      score, specified by <code>--gap-cost</code>.
      </li><li><code>wildcard</code> The gap character matches any base, and the
      score is the product of the corresponding probabilities.
      </li><li><code>minimum</code> The gap character is assigned the score
      corresponding to the least likely letter at the given position.
      </li><li><code>model</code> Use model-specific gap handling. Currently, the only
      model that supports this is <code>f81_gap</code>.
      </li></ul></li>
      <li> <code>--gap-cost <i>&lt;float&gt;</i></code>
    - Specifies the costs for gaps when
      using the <code>fixed</code> gap handling strategy. Default is 0.0.
  </li>
      <li> <code>--hb</code> 
    - Use the Halpern-Bruno modification to the evolutionary model.
  </li>
      <li> <code>--list</code>
    - Treat the second required input as a list of
    alignments, rather than a single alignment.
  </li>
      <li> <code>--max-stored-scores <code><i>&lt;max&gt;</i></code></code>
    - Set the maximum number of scores that will be stored.
    Precise calculation of q-values depends on having a complete list of scores.
    However, keeping a complete list of scores may exceed available memory.
    Once the number of stored scores reaches the maximum allowed, 
    the least significant 50% of scores will be dropped,
    and approximate q-values will be calculated.
    By default the maximum number of stored matches is 100,000.
  </li>
      <li> <code>--model [single|average|jc|k2|f81|f84|hky|tn]</code>
    - The evolutionary model to use.
    The available models are:
    <ul><li>single - score first sequence: compute standard log-odds score of
	  first sequence in the alignment; ignores tree but does NOT remove gaps.</li><li>average - compute average of standard log-odds score of aligned sites.</li><li>jc - Jukes-Cantor: equilibrium base frequencies are all 1/4;
    the only free parameter is the mutation rate.</li><li>k2 - Kimura 2-parameter: equilibrium base frequencies are all 1/4;
    the free parameters are the mutation rate and the transition/transversion rate ratio.</li><li>f81 - Felsenstein 1981: equilibrium base frequencies are taken from the alignment;
    the only free parameter is the mutation rate.</li><li>f84 - Felsenstein 1984: equilibrium base frequencies are taken
    from the alignment; the free parameters are the mutation rate and the
    transition/transversion rate ratio.  The ratio of purine-purine to
    pyrimidine-&gt;pyrimidine transitions is assumed to be 1.</li><li>hky - Hasegawa-Kishino-Yano: equilibrium base frequencies are
    taken from the alignment; the free parameters are the mutation rate
    and the transition/transversion rate ratio.  The ratio of
    purine-purine to pyrimidine-pyrimidine transitions is assumed to be
    equal to the  ratio of purines to pyrimidines.</li><li>tn - Tamura-Nei: equilibrium base frequencies are taken from the
    alignment;  the free parameters are the mutation rate, the
    transition/transversion rate ratio, and the ratio of purine-purine
    transitions to pyrimidine-pyrimidine transitions.</li></ul><p>
    The default model is f81.  A description of the f81 model is available
    in chapter 13 of <i>Statistical Methods in Bioinformatics</i> by Ewens
    and Grant. The other models are described in chapters 9 and 13 of
    <i>Inferring Phylogenies</i> by Felsenstein.
    </p></li>
      <li> <code>--motif <i>&lt;id&gt;</i></code>
    - Use only the motif identified by <code><i>&lt;id&gt;</i></code>. 
    This option may be repeated.
  </li>
      <li> <code>--no-qvalue</code>
    - Do not compute a q-value for each p-value.  The q-value 
    calculation is that of Benjamini and Hochberg (1995).
    By default, q-values are computed.
  </li>
      <li> <code>--norc</code>
    - Do not score the reverse complement DNA strand.  Both
	strands are scored by default.
  </li>
      <li> <code>--o <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will not be overwritten.
  </li>
      <li> <code>--oc <i>&lt;dir name&gt;</i></code>
    - Specifies the output directory.
    If the directory already exists, the contents will be overwritten.
  </li>
      <li> <code>--output-pthresh <i>&lt;float&gt;</i></code>
    - The p-value threshold for displaying search results.
    If the p-value of a match is greater than this value,
    then the match will not be printed. 
    If both the <code>--output-pthresh</code> and <code>--output-qthresh</code>
    options appear on the command line, whichever appears later
    on the command line will be applied.
    The default p-value threshold is 1e-4.
  </li>
      <li> <code>--output-qthresh <i>&lt;float&gt;</i></code>
    - The q-value threshold for displaying search results.
    If the q-value of a match is greater than this value,
    then the match will not be printed.
    By default, no q-value threshold is applied.
  </li>
      <li> <code>--pseudocounts &lt;float&gt;</code> 
    - A pseudocount to be added to each count in the motif matrix, 
    weighted by the background frequencies for the nucleotides (Dirichlet prior), 
    before converting the motif to probabilities. 
    The default value is 0.1.
  </li>
      <li> <code>--pur-pyr &lt;float&gt;</code>
    - The ratio of the purine transition rate to pyrimidine transition rate.
    This parameter is used by the Tamura-nei model.
    The default value is 1.0.
  </li>
      <li> <code>--seed &lt;n&gt;</code>
    - Seed for random number generator.
  </li>
      <li> <code>--text</code>
    Limits output to plain text sent to standard out.  For FIMO, the
    text output is unsorted, and q-values are not reported.  This mode
    allows the program to search an arbitrarily large database,
    because results are not stored in memory.
  </li>
      <li> <code>--transition-transversion &lt;float&gt;</code>
    - The ratio of the transition rate to the transversion rate.
    This parameter is used by the Kimura 2-parameter, F84, HKY,
    and Tamura-nei models.
    The default value is 0.5.
  </li>
      <li> <code>--verbosity 1|2|3|4</code>
    - Set the verbosity of status reports to standard error.
    The default level is 2.
  </li>
      </ul>
      <b>Warning messages:</b> None

      <b>Bugs:</b>

      <ul>

      <li>
      F84 and Tamura-Nei models fail trying to take the log of a negative number.</li>
      <li>
      Allow local realignment, a la Monkey.</li>
      <li>
      The <code>--motif</code> option should allow multiple motifs
      to be selected from the motif file.
      </li>
      <li>
      Print motif consensus as part of feature properties.</li>
      <li>
      Allow reading of a column frequency file; this requires
      writing a stand-alone to create such a file and establishing
      a format such as:
      <pre>
         # species_list_1
         AAA f_AAA
         ...
         TTT f_TTT
         # species_list_2
         AAAAA f_AAAAA
         ...
         TTTTT f_TTTTT
         ...
      </pre>
      </li>

      </ul>

      </blockquote>
    </div>
  </body>
</html>
