<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <title>SpaMo - motif spacing analysis tool</title>
    <link href="meme-suite-docs.css" rel="stylesheet" type="text/css">
    <link rel="icon" type="image/png" href="../images/memesuite_icon.png">
  </head>
  <body>
    <div class="memesuitelogo">
      <h1>The MEME Suite</h1>
      <h2>Motif-based sequence analysis tools</h2>
    </div>
    <h2><img src="images/spamo_logo.png" alt="SpaMo - Spaced Motif analysis tool"
      title="Spaced Motif analysis tool"></h2>
    <div class="usage">
      <h3>Usage:</h3>
      <code>spamo [options] &lt;sequences&gt; &lt;primary motif&gt; &lt;secondary motifs&gt;+</code>
    </div>
    <h3>Description</h3>
    <div class="indent">
      <p>SpaMo does inference of transcription factor complexes by looking for
      significant spacings between binding sites.</p>

      <p>The inputs are a set of many short sequences, a primary motif and one
      or more databases of secondary motifs. It searches for the strongest
      primary motif binding site and then searches in the area around it looking
      for the strongest secondary motif binding site. The relative spacings of
      the primary and secondary motif in all the sequences is tallied and the
      probability of the close spacings happening by chance is calculated.</p>

      <p>After all the calculations are done SpaMo outputs the non-redundant
      secondary motifs in order of significance provided they had a bin that
      passed the significance threshold. Similar secondary motifs are grouped
      together and listed in order of significance on the secondary motif they
      were redundant to. If the bin size is one then an alignment for each of
      the similar secondary motifs is created.</p>
    </div>

    <h3>Inputs</h3>
    <div class="indent">
      <h4>Sequences</h4>
      <div class="indent">
        <p>A FASTA formatted file containing lots of short sequences centered on
        a site expected to be relevant to the primary motif. This would
        typically be generated by expanding either side of a ChIP-seq peak to
        obtain sequences of about 500 bases in length.</p>
        <p>SpaMo scans the central section, excluding the margin on either edge,
        for the primary motif. As the margin on each edge is excluded then if
        the sequence is shorter than two times the margin plus the trimmed
        length of the primary motif the sequence will always be discarded.</p>
      </div>

      <h4>Primary Motif</h4>
      <div class="indent">
        <p>A MEME formatted motif file containing a DNA motif. The primary motif
        is the motif for which you are trying to find cofactors. If the file
        contains more than one motif then the first will be selected by default
        or another can be selected using the -primary or -primaryi options.</p>
      </div>

      <h4>Secondary Motifs</h4>
      <div class="indent">
        <p>One or more MEME formatted motif files containing DNA motifs. The
        secondary motifs are tested for a significant spacing with the primary
        motif which might imply they act together. If the motif databases
        contain motifs which you don't wish to scan, the motifs can be filtered
        based on their name by using the -inc and -exc options.</p>
      </div>
    </div>

    <h3>Optional Inputs</h3>
    <div class="indent">
      <h4>Motif CISMLs</h4>
      <div class="indent">
        <p>SpaMo can determine the best primary and secondary motif sites from scores in CISML files.</p>
      </div>
      <h4>Motif Background</h4>
    </div>

    <h3>Outputs</h3>
    <div class="indent">
      <p>SpaMo outputs its output to files in a directory named
      <code>spamo_out</code>, which it creates if necessary. You can change the
      output directory using the -o or -oc options.</p> 
      
      <p>The main output file is named <code>spamo.html</code> and can be viewed
      with a web browser. The <code>spamo.html</code> file is generated from the
      <code>spamo.xml</code> file so using the xml file is recommended when
      machine processing is required.</p> 
      
      <p>The histograms are only generated when the -eps and/or the -png options
      are specified. If you are viewing the output in older web-browsers you
      will need to specify the -png option so the histograms are viewable.</p>
    </div>
    <h3>Options</h3>
    <div class="indent">
    <table class="options">
      <thead>
        <tr>
          <th>Option</th>
          <th>Parameter</th>
          <th>Description</th>
          <th>Default Behaviour</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th colspan="4">Input/Output</th>
        </tr>
        <tr>
          <td>-o</td>
          <td><span class="pdat">directory</span></td>
          <td>Create the directory and write output files in it. This option is
          not compatible with -oc as only one output directory is allowed.</td>
          <td>The program behaves as if <code>-oc&nbsp;spamo_out</code> had
          been specified.</td>
        </tr>
        <tr>
          <td>-oc</td>
          <td><span class="pdat">directory</span></td>
          <td>Create the directory but if it already exists allow overwriting
          the contents. This option is not compatible with -o as only one output
          directory is allowed.</td>
          <td>The program behaves as if <code>-oc&nbsp;spamo_out</code> had been
          specified.</td>
        </tr>
        <tr>
          <td>-loadcismls</td>
          <td>&nbsp;</td>
          <td>Load CISML files to get motif position scores instead of scanning.
            If this flag is specified then each motif file must have a CISML
            file specified after it. This is not compatible with -trim as that
            option must modify the motifs before scanning.</td>
          <td>Scan sequences to determine the position scores.</td>
        </tr>
        <tr>
          <td>-eps</td>
          <td>&nbsp;</td>
          <td>Output histograms in Encapsulated PostScript format which can be
          included in publications. This option can be used with the -png
          option.</td>
          <td>Image files are not output by default as the webpage is capable of
          generating the graphs on demand.</td>
        </tr>
        <tr>
          <td>-png</td>
          <td>&nbsp;</td>
          <td>Output histograms in Portable Network Graphic format which is good
          for webpages. This option can be used with the -eps option</td>
          <td>Image files are not output by default as the webpage is capable of
          generating the graphs on demand.</td>
        </tr>
        <tr>
          <td>-dumpseqs</td>
          <td>&nbsp;</td>
          <td>Write space separated values in columns, describing the motif 
            matches used to make the histograms, to output files. The rows are 
            initially in sequence name order but various command-line tools can
            be used to sort them on other values. The columns contain:
<table>
  <thead>
    <tr><th>column(s)</th><th>contents</th></tr>
  </thead>
  <tbody>
    <tr><td>1</td><td>Trimmed lowercase sequence with uppercase
        matches</td></tr>
    <tr><td>2</td><td>Position of the secondary match within the whole
        sequence</td></tr>
    <tr><td>3</td><td>Sequence fragment that the primary matched</td></tr>
    <tr><td>4</td><td>Strand of the primary match (+/-)</td></tr>
    <tr><td>5</td><td>Sequence fragment that the secondary matched</td></tr>
    <tr><td>6</td><td>Strand of the secondary match (+/-)</td></tr>
    <tr><td>7</td><td>Is the primary match on the same strand as the
        secondary (s/o)</td></tr>
    <tr><td>8</td><td>Is the secondary match downstream or
        upstream (d/u)</td></tr>
    <tr><td>9</td><td>The gap between the primary and secondary
        matches</td></tr>
    <tr><td>10</td><td>The name of the sequence</td></tr>
    <tr><td>11</td><td>The <i>p</i>-value of the bin containing the match, 
        adjusted for the number of bins</td></tr>
    <tr><th colspan="2">If the sequence names are in Genome Browser position
    format (e.g., "chr5:36715616-36715623"), the following additional columns
    appear:</th></tr>
    <tr><td>12-14</td><td>Position of primary match in BED coordinates</td></tr>
    <tr><td>15</td><td>Position of primary match in Genome Browser
        coordinates</td></tr>
    <tr><td>16-18</td><td>Position of secondary match in BED
        coordinates</td></tr>
    <tr><td>19</td><td>Position of secondary match in Genome Browser
        coordinates</td></tr>
  </tbody>
</table>
          </td>
          <td>No specific match information is output.</td>
        </tr>
        <tr>
          <td>-dumpsigs</td>
          <td>&nbsp;</td>
          <td>Same as -dumpseqs, but only secondary matches in significant
          bins are dumped.</td>
          <td>As in -dumpseqs.</td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th colspan="4">Scanning</th>
        </tr>
        <tr>
          <td>-numgen</td>
          <td><span class="pdat">seed</span></td>
          <td>Specify a number as the seed for initializing the pseudo-random
          number generator used in breaking scoring ties. The seed is included
          in the output so experiments can be repeated. If you wish to run
          multiple experiments with different seeds then you can use the special
          value 'time' (without the quotes) which sets the seed to the system
          clock.</td>
          <td>A seed of 1 is used.</td>
        </tr>
        <tr>
          <td>-margin</td> 
          <td><span class="pdat">size</span></td>   
          <td>The distance either side of the primary motif site which makes up
          the region that can contain the secondary motif site. Additionally it
          is the minimum gap between the primary motif site and the edge of the
          sequence. These constraints mean that input sequences shorter than
          the trimmed length of the primary motif plus two times the margin size
          can not be used by SpaMo.</td> 
          <td>A margin of 150 is used. For an input sequence of length 500 this 
          means the central 200 bases are scanned for the best primary motif 
          match and then the 300 bases surrounding the best primary site are
          scanned for the best secondary site.</td>
        </tr>
        <tr>
          <td>-bin</td>
          <td><span class="pdat">size</span></td>
          <td>The size of the bin used to calculate the histogram and
          <i>p</i>-values. A bin size of 1 is recommended as it gives better
          output.</td>
          <td>A bin size of 1 is used.</td>
        </tr>
        <tr>
          <td>-range</td>
          <td><span class="pdat">size</span></td>
          <td>The distance from the primary motif site for which <i>p</i>-values
          are calculated to include in significance tests. A small value for
          range may miss significant peaks but this is a trade-off as a the
          larger the range the more bins have to be tested leading to a larger
          factor used in the Bonferroni correction for multiple tests.</td>
          <td>A range of 150 is used.</td>
        </tr>
        <tr>
          <td>-shared</td>
          <td><span class="pdat">fraction</span></td>
          <td>After the primary motif site has been selected in each sequence
          the sequence is trimmed to only include a region of size margin on
          either side of the primary motif site. This aligned and trimmed
          sequence is then compared with all the other sequences and the
          fraction of shared bases is calculated. If the fraction of shared
          bases is larger than this limit then one of the sequences is
          eliminated. To disable this feature set the shared fraction to 1.</td>
          <td>The shared fraction is set to 0.5 which means that the trimmed,
          aligned sequences must share 50% or more of their bases to be declared
          redundant.</td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th colspan="4">Summarizing</th>
        </tr>
        <tr>
          <td>-cutoff</td>
          <td><span class="pdat"><i>p</i>-value</span></td>
          <td>The <i>p</i>-value cutoff for bins to be considered significant. 
          This is the <i>p</i>-value of the Binomial Test on the number of
          observed secondary spacings or more falling into the given bin,
          adjusted for the number of bins tested. Note that the <i>p</i>-value
          is only calculated and tested for bins within the distance of the
          primary motif as specified by the option -range.</td>
          <td>A bin <i>p</i>-value smaller than or equal to 0.05 is considered
          significant.</td>
        </tr>
        <tr>
          <td>-evalue</td>
          <td><span class="pdat"><i>E</i>-value</span></td>
          <td>The minimum secondary motif <i>E</i>-value for its results to be
          printed. For each secondary motif, this is the minimum <i>p</i>-value
          of all tested bins multipled by the number of secondary motifs. The
          <i>E</i>-value estimates the expected number of random secondary
          motifs that would have the given <i>E</i>-value or lower.</td>
          <td>Results for all secondary motifs with <i>E</i>-value smaller than
          or equal to 10 are printed.</td>
        </tr>
        <tr>
          <td>-overlap</td>
          <td><span class="pdat">size</span></td>
          <td>To determine if two motifs are redundant the most significant bin
          in the tested range in each of the motifs is compared. For the motifs
          to be considered redundant it needs to be possible that the sites that
          got counted in the bin could have overlapped, and this parameter sets
          the minimum overlap. For a bin size larger than 1 the overlap of the
          bins can not be precisely calculated as the actual site positions are
          not stored and so the maximum possible overlap is used.</td>
          <td>A minimum overlap of 2 is required.</td>
        </tr>
        <tr>
          <td>-joint</td> 
          <td><span class="pdat">fraction</span></td>
          <td>To determine if two motifs are redundant the most significant bin
          in the tested range in each of the motifs is compared. The most
          significant bin in each motif has the list of sequence identifiers
          which had a primary and secondary at the correct spacing to go into
          that bin. To compare the motifs for redundancy this set of sequence
          identifiers is compared and the size of the intersection is counted.
          This intersection size is divided by the size of the smaller of the
          two sequence sets to get the joint sequence fraction.</td>
          <td>A minimum joint sequence fraction of 0.5 is required for two
          motifs to be considered redundant.</td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th colspan="4">Motif Loading</th>
        </tr>
        <tr>
          <td>-pseudo</td>
          <td><span class="pdat">count</span></td>
          <td>The pseudocount added to loaded motifs.</td>
          <td>A pseudocount of 0.1 is added to loaded motifs.</td>
        </tr>
        <tr>
          <td>-bgfile</td><td><span class="pdat">file</span></td>
          <td>The file containing the background frequency information used in
          applying pseudocounts.</td>
          <td>The frequencies of bases in the sequences are used as a
          background.</td>
        </tr>
        <tr>
          <td>-trim</td><td><span class="pdat">bits</span></td>
          <td>Trim the edges of motifs based on the information content. The
          positions on the edges of the motifs with information content less
          than bits will not be used in scanning. This is incompatible with the
          -loadcismls option as the motifs must be trimmed before scoring can
          take place.</td>
          <td>Positions on the edges of the motifs with information content
          less than or equal to 0.25 will be trimmed.</td>
        </tr>
        <tr>
          <td>-primary</td><td><span class="pdat">name</span></td>
          <td>The name of the motif to select as the primary motif. This option
          is incompatible with -primaryi as only one primary motif can be
          selected.</td>
          <td>The first motif in the file is selected.</td>
        </tr>
        <tr>
          <td>-primaryi</td><td><span class="pdat">num</span></td>
          <td>The index of the motif to select as the primary motif counting
          from 1. This option is incompatible with -primary as only one primary
          motif can be selected.</td>
          <td>The first motif in the file is selected.</td>
        </tr>
        <tr>
          <td>-keepprimary</td><td>&nbsp;</td>
          <td>If the same file is specified for the primary and secondary motifs
          then by default the primary motif is excluded but specifying this
          option keeps it.</td>
          <td>The primary motif is excluded from the secondaries if the same
          file is used for the primary and secondary motifs.</td>
        </tr>
        <tr>
          <td>-inc</td><td><span class="pdat">pattern</span></td>
          <td>Select the motifs with names matching the pattern. The pattern can
          contain shell like wildcards (e.g., '*') though they must be escaped
          or quoted to prevent the shell from auto-expanding them. This option
          may be may be repeated and all the patterns will be used.</td> 
          <td>Unless the -exc option has been specified all the motifs are
          used.</td>
        </tr>
        <tr>
          <td>-exc</td><td><span class="pdat">pattern</span></td>
          <td>Exclude the motifs with names matching the pattern. The pattern
          can contain shell like wildcards (e.g., '*') though they must be
          escaped or quoted to prevent the shell from auto-expanding them. This
          option may be may be repeated and all the patterns will be used.</td> 
          <td>Unless the -inc option has been specified all the motifs are
          used.</td>
        </tr>
      </tbody>
      <tbody>
        <tr>
          <th colspan="4">Miscellaneous</th>
        </tr>
        <tr>
          <td>-help</td><td>&nbsp;</td>
          <td>Print out a help message.</td>
          <td>&nbsp;</td>
        </tr>
        <tr>
          <td>-verbosity</td><td>1|2|3|4|5</td>
          <td>A number that regulates the verbosity level of the output
          information messages. If set to 1 (quiet) then it will only output
          error messages whereas the other extreme 5 (dump) outputs lots of
          mostly useless information.</td>
          <td>The verbosity level is set to 2 (normal).</td>
        </tr>
      </tbody>
    </table>
    </div>
  </body>
</html>
